File: src/engine/search/link.py
"""
KIVO Search Engine
Link

Zwei Arten: manuell (vom Nutzer gesetzt) und suggested (automatisch berechnet).
Reihenfolge ist immer: erst manuelle Links, dann Vorschlaege. Maximal 3 gesamt.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal
from uuid import UUID

LinkKind = Literal["manual", "suggested"]


@dataclass(slots=True, frozen=True)
class Link:
    target_id: UUID
    kind: LinkKind
    score: float = 0.0  # nur relevant fuer "suggested", zum Ranking


def combine_links(manual: list[Link], suggested: list[Link], max_total: int = 3) -> list[Link]:
    """
    Regel aus der Spec:
      0 manuelle  -> 3 Vorschlaege
      1 manuelle  -> 1 manuell + 2 Vorschlaege
      2 manuelle  -> 2 manuell + 1 Vorschlag
      3+ manuelle -> nur die manuellen, 0 Vorschlaege
    Manuelle Links kommen immer zuerst.
    """
    manual = manual[:max_total]
    remaining_slots = max(0, max_total - len(manual))

    # Vorschlaege sortiert nach Score (bester zuerst), Duplikate zu manuellen raus
    manual_targets = {l.target_id for l in manual}
    ranked_suggestions = sorted(
        (s for s in suggested if s.target_id not in manual_targets),
        key=lambda s: s.score,
        reverse=True,
    )

    return manual + ranked_suggestions[:remaining_slots]

File: src/frontend/src-tauri/src/engine/link.rs
//! KIVO Search Engine (Rust-Port)
//! Link
//!
//! 1:1-Port von `src/engine/search/link.py`.
//!
//! STRUKTUR-HINWEIS: `Link`/`LinkKind` selbst liegen bei diesem Port
//! bereits seit Phase 1 in `entry.rs` (weil `Entry` sie direkt braucht) -
//! im Python-Original ist das getrennt (`entry.py` importiert aus
//! `link.py`). Rein strukturelle Abweichung, keine Verhaltensaenderung.
//! Diese Datei enthaelt nur die eigentliche Kombinier-Logik.
//!
//! Regel aus dem Original: 0 manuelle -> 3 Vorschlaege, 1 manuelle -> 1
//! manuell + 2 Vorschlaege, 2 manuelle -> 2 manuell + 1 Vorschlag, 3+
//! manuelle -> nur die manuellen, 0 Vorschlaege. Manuelle Links kommen
//! immer zuerst.

use std::cmp::Ordering;
use std::collections::HashSet;
use uuid::Uuid;

use crate::engine::entry::Link;

/// Entspricht `combine_links(manual, suggested, max_total)`.
pub fn combine_links(manual: &[Link], suggested: &[Link], max_total: usize) -> Vec<Link> {
    let manual: Vec<Link> = manual.iter().take(max_total).cloned().collect();
    let remaining_slots = max_total.saturating_sub(manual.len());

    let manual_targets: HashSet<Uuid> = manual.iter().map(|l| l.target_id).collect();

    let mut ranked_suggestions: Vec<Link> = suggested
        .iter()
        .filter(|s| !manual_targets.contains(&s.target_id))
        .cloned()
        .collect();

    // Python: `sorted(..., key=lambda s: s.score, reverse=True)` (stabil).
    ranked_suggestions.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap_or(Ordering::Equal));

    let mut result = manual;
    result.extend(ranked_suggestions.into_iter().take(remaining_slots));
    result
}
